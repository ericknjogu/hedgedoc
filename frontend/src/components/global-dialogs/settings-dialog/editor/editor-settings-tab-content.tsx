/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */
import { SettingLine } from '../utils/setting-line'
import { LigatureSettingButtonGroup } from './ligature-setting-button-group'
import { LineWrappingSettingButtonGroup } from './line-wrapping-setting-button-group'
import { SmartPasteSettingButtonGroup } from './smart-paste-setting-button-group'
import { SyncScrollSettingButtonGroup } from './sync-scroll-setting-button-group'
import React from 'react'
import { ListGroup } from 'react-bootstrap'
import { useTranslation } from 'react-i18next'
import { SpellcheckSettingButtonGroup } from './spellcheck-setting-button-group'
import { IndentWithTabsSettingButtonGroup } from './indent-with-tabs-setting-button-group'
import { useApplicationState } from '../../../../hooks/common/use-application-state'
import { IndentSpacesSettingInput } from './indent-spaces-setting-input'

/**
 * Shows the editor specific settings.
 */
export const EditorSettingsTabContent: React.FC = () => {
  useTranslation()
  const useTabs = useApplicationState((state) => state.editorConfig.indentWithTabs)

  return (
    <ListGroup>
      <SettingLine i18nKey={'editor.ligatures'}>
        <LigatureSettingButtonGroup />
      </SettingLine>
      <SettingLine i18nKey={'editor.smartPaste'}>
        <SmartPasteSettingButtonGroup />
      </SettingLine>
      <SettingLine i18nKey={'editor.syncScroll'}>
        <SyncScrollSettingButtonGroup />
      </SettingLine>
      <SettingLine i18nKey={'editor.lineWrapping'}>
        <LineWrappingSettingButtonGroup />
      </SettingLine>
      <SettingLine i18nKey={'editor.spellCheck'}>
        <SpellcheckSettingButtonGroup />
      </SettingLine>
      <SettingLine i18nKey={'editor.indentWithTabs'}>
        <IndentWithTabsSettingButtonGroup />
      </SettingLine>
      {!useTabs && (
        <SettingLine i18nKey={'editor.indentSpaces'}>
          <IndentSpacesSettingInput />
        </SettingLine>
      )}
    </ListGroup>
  )
}
