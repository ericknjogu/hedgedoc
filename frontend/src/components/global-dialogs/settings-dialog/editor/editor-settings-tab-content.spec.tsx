/*
 * SPDX-FileCopyrightText: 2023 The HedgeDoc developers (see AUTHORS file)
 *
 * SPDX-License-Identifier: AGPL-3.0-only
 */

import { mockAppState } from '../../../../test-utils/mock-app-state'
import { render } from '@testing-library/react'
import { EditorSettingsTabContent } from './editor-settings-tab-content'
import { mockI18n } from '../../../../test-utils/mock-i18n'

jest.mock('../../../../hooks/common/use-application-state')

describe('EditorSettingsTabContent', () => {
  beforeEach(async () => {
    await mockI18n()
  })

  afterEach(() => {
    jest.resetAllMocks()
    jest.resetModules()
  })

  it('renders space settings when indentWithTabs is false', () => {
    mockAppState({
      editorConfig: {
        syncScroll: false,
        spellCheck: false,
        smartPaste: false,
        lineWrapping: false,
        ligatures: false,
        indentWithTabs: false,
        indentSpaces: 7
      }
    })
    const view = render(<EditorSettingsTabContent />)
    expect(view.container).toMatchSnapshot()
  })

  it('hides space settings indentWithTabs is true', () => {
    mockAppState({
      editorConfig: {
        syncScroll: false,
        spellCheck: false,
        smartPaste: false,
        lineWrapping: false,
        ligatures: false,
        indentWithTabs: true,
        indentSpaces: 7
      }
    })
    const view = render(<EditorSettingsTabContent />)
    expect(view.container).toMatchSnapshot()
  })
})
