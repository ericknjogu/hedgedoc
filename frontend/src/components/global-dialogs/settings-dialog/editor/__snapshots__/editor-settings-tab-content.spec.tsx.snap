// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditorSettingsTabContent hides space settings indentWithTabs is true 1`] = `
<div>
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.ligatures.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-ligatures"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-ligatures"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.ligatures.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.smartPaste.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-smart-paste"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-smart-paste"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.smartPaste.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.syncScroll.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-sync-scroll"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-sync-scroll"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.syncScroll.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.lineWrapping.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-line-wrapping"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-line-wrapping"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.lineWrapping.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.spellCheck.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-spell-check"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-spell-check"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.spellCheck.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.indentWithTabs.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-indent-with-tabs"
              title="settings.editor.indentWithTabs.tabs"
              type="radio"
            >
              settings.editor.indentWithTabs.tabs
            </button>
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-indent-with-tabs"
              title="settings.editor.indentWithTabs.spaces"
              type="radio"
            >
              settings.editor.indentWithTabs.spaces
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.indentWithTabs.help
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`EditorSettingsTabContent renders space settings when indentWithTabs is false 1`] = `
<div>
  <div
    class="list-group"
  >
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.ligatures.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-ligatures"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-ligatures"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.ligatures.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.smartPaste.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-smart-paste"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-smart-paste"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.smartPaste.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.syncScroll.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-sync-scroll"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-sync-scroll"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.syncScroll.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.lineWrapping.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-line-wrapping"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-line-wrapping"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.lineWrapping.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.spellCheck.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-spell-check"
              title="common.on"
              type="radio"
            >
              common.on
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-spell-check"
              title="common.off"
              type="radio"
            >
              common.off
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.spellCheck.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.indentWithTabs.label
        </div>
        <div
          class="col-md-4"
        >
          <div
            class="btn-group"
            role="group"
          >
            <button
              class="btn btn-outline-secondary"
              data-testid="onOffButtonGroupOn"
              name="settings-indent-with-tabs"
              title="settings.editor.indentWithTabs.tabs"
              type="radio"
            >
              settings.editor.indentWithTabs.tabs
            </button>
            <button
              class="btn btn-secondary"
              data-testid="onOffButtonGroupOff"
              name="settings-indent-with-tabs"
              title="settings.editor.indentWithTabs.spaces"
              type="radio"
            >
              settings.editor.indentWithTabs.spaces
            </button>
          </div>
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.indentWithTabs.help
        </div>
      </div>
    </div>
    <div
      class="list-group-item"
    >
      <div
        class="row"
      >
        <div
          class="col-md-3"
        >
          settings.editor.indentSpaces.label
        </div>
        <div
          class="col-md-4"
        >
          <input
            class="w-auto form-control"
            min="1"
            type="number"
            value="7"
          />
        </div>
        <div
          class="col-md-5"
        >
          settings.editor.indentSpaces.help
        </div>
      </div>
    </div>
  </div>
</div>
`;
